lib/hawkbit/hawkbit.c

```diff
@@ -0,0 +1,1261 @@
+	LOG_DBG("chunks[0].version=%s", log_strdup(c->version));
+	LOG_DBG("chunks[0].artifacts[0].filename=%s",
+		log_strdup(a->filename));
+	LOG_DBG("chunks[0].artifacts[0].hashes.sha1=%s",
+		log_strdup(a->hashes.sha1));
+	LOG_DBG("chunks[0].artifacts[0].hashes.md5=%s",
+		log_strdup(a->hashes.md5));
+	LOG_DBG("chunks[0].artifacts[0].hashes.sha256=%s",
+		log_strdup(a->hashes.sha256));
+	LOG_DBG("chunks[0].size=%d", a->size);
+	LOG_DBG("download-http=%s",
+		log_strdup(l->download_http.href));
+	LOG_DBG("md5sum =%s", log_strdup(l->md5sum_http.href));
+}
+
+int hawkbit_init(void)
+{
+	bool image_ok;
+	int ret = 0, rc = 0;
+	struct flash_pages_info info;
+	int32_t action_id;
+
+	fs.offset = FLASH_AREA_OFFSET(storage);
+	rc = flash_get_page_info_by_offs(
+		device_get_binding(DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL),
+		fs.offset, &info);
+	if (rc) {
+		LOG_ERR("Unable to get storage page info");
+		return -EIO;
+	}
+
+	fs.sector_size = info.size;
+	fs.sector_count = 3U;
+
+	rc = nvs_init(&fs, DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL);
+	if (rc) {
+		LOG_ERR("Storage flash Init failed");
+		return -ENODEV;
+	}
+
+	rc = nvs_read(&fs, ADDRESS_ID, &action_id, sizeof(action_id));
+	LOG_DBG("Action id: current %d", action_id);
+
+	image_ok = boot_is_img_confirmed();
+	LOG_INF("Image is %s confirmed OK", image_ok ? "" : " not");
+	if (!image_ok) {
+		ret = boot_write_img_confirmed();
+		if (ret < 0) {
+			LOG_ERR("Couldn't confirm this image: %d", ret);
+			return ret;
+		}
+
+		LOG_DBG("Marked image as OK");
+		ret = boot_erase_img_bank(FLASH_AREA_ID(image_1));
+		if (ret) {
+			LOG_ERR("Failed to erase second slot");
+			return ret;
+		}
+	}
+
+	return ret;
+}
+
+static int enum_for_http_req_string(char *userdata)
+{
+	int i = 0;
+	char *name = http_request[i].http_req_str;
+
+	while (name) {
+		if (strcmp(name, userdata) == 0) {
+			return http_request[i].n;
+		}
+
+		name = http_request[++i].http_req_str;
+	}
+
+	return 0;
+}
+
+static void response_cb(struct http_response *rsp,
+			enum http_final_call final_data,
+			void *userdata)
+{
+	static size_t body_len;
+	int ret, type, downloaded;
+	uint8_t *body_data = NULL;
+	static size_t response_buffer_size = RESPONSE_BUFFER_SIZE;
+
+	type = enum_for_http_req_string(userdata);
+
+	switch (type) {
+	case HAWKBIT_PROBE:
+		if (hb_context.dl.http_content_size == 0) {
+			body_data = rsp->body_start;
+			body_len = strlen(rsp->recv_buf);
+			body_len -= (rsp->body_start - rsp->recv_buf);
+			strncpy(hb_context.response_data, body_data, body_len);
+			hb_context.dl.http_content_size = rsp->content_length;
+		}
+
+		if ((body_data == NULL) && (final_data == HTTP_DATA_MORE)) {
+			body_data = rsp->recv_buf;
+			body_len = body_len + (rsp->data_len);
+			if (body_len > response_buffer_size) {
+				response_buffer_size <<= 1;
+				hb_context.response_data =
+					realloc(hb_context.response_data,
+						response_buffer_size);
+				if (hb_context.response_data == NULL) {
```

Comment:
We will be leaking original content of `hb_context.response_data` if `realloc()` returns NULL. You need to use a temp variable here.

Reply:
If memory reallocation failed, there is no use with original content. Again hawkbit client need to pull the content from server.

Reply:
Do you mean that `hb_context.response_data` is always NULL before the realloc() call? If that is the case, then just use malloc() here.
If `hb_context.response_data` contains something, and realloc() returns NULL, then we leak the original value of hb_context.response_data.  The manual page of realloc says `If realloc() fails, the original block is left untouched; it is not  freed or moved.`